Make ctrl-up/down change speed modifier of mods

LeaUndercover opened the pull request:
PageUp and PageDown now increase or decrease speed in song select.

**Example**
(Song Select screen and no mod is selected)
1. Page Down is pressed
2. HalfTime mod gets applied with 0.95x speed modifier
3. Page Down is pressed again
4. HalfTime mod speed modifier changes to 0.9x
same goes with PageUp and DoubleTime.

If Daycore or Nightcore are currently selected it increases those.
If the minimum rate gets surpassed Half-/Double-Time get applied

Video of it in action: 
https://github.com/ppy/osu/assets/28962347/e95228d2-0e77-4ec8-9e60-17472bdc4545

Im open to any suggestions for changes and i hope this feature finds its way into osu!lazer

Commit:
Initial implementation

Commit:
F1 also does not work with minus in song select, same behaviour

Commit:
Comment, make code more readable, functions are now private

LeaUndercover committed:
Merge branch 'master' into osu-lazer-speedkeys

LeaUndercover committed:
Merge pull request #1 from Fabiano1337/osu-lazer-speedkeys

Osu lazer speedkeys

Commit:
Fix Formatting

Commit:
Fix Code Inspection

bdach commented:
- What is the reason for wanting this?
- Why should DT/HT get preferential hotkey treatment over every other mod in the game?
- Was this feature discussed anywhere at all before this PR?

Fabiano1337 commented:
**What is the reason for wanting this?**
It is a quality of life feature that makes adjusting the speed of your map faster (useful for people who often push their limits and increase the speed every run of a map), with this implementation you can adjust speed the using only your keyboard.

**Why should DT/HT get preferential hotkey treatment over every other mod in the game?**
DT/HT are some of the most used mods in the game, HR/HD don't have any adjustability. HR and HD can be fully utilized with only a keyboard whereas DT/HT cannot. DC/NC are lesser used than DT/HT which is why i opted for DT/HT, this is up for debate tho.

**Was this feature discussed anywhere at all before this PR?**
No, i haven't seen this feature being discussed anywhere before, since isn't that big of a feature i thought it didn't need one. Do you want me to open a Discussion about this topic first ?

longnguyen2004 commented:
(Not lazer dev, just my opinion) The hotkey for this cannot be PgUp/PgDown, since those are used to scroll the carousel, and I also don't like the semantic of PgUp/PgDown being changed like that, since those keys are almost always used to scroll, and not to adjust a setting. Maybe Ctrl+Up and Ctrl+Down?

bdach commented:
> No, i haven't seen this feature being discussed anywhere before, since isn't that big of a feature i thought it didn't need one. Do you want me to open a Discussion about this topic first ?

I generally prefer all unsolicited feature requests to go through discussions first in order to reduce wasted contributor effort and to gauge interest in the feature. At this time I have no idea as to whether anyone but yourself wants this feature to exist.

---

UX-wise I have a whole mess of doubts.

- As stated above this should not use page up/down by default.
- If you manually select daycore and then page-up enough to progress into track speed-up I'd probably expect nightcore to be activated. Right now double time activates.
- If you manually change "adjust pitch" on DT/HT and then page-up/down enough to progress into HT/DT then mod settings are lost. They should probably be preserved.
- It's not very clear what's happening if you press the binding but don't look at the footer. This gets especially weird if you e.g. speed up DT from 1.45x to 1.5x because the speed indicator on the right of the right of the mod icon in the footer disappears. I'd say there should be a toast or something showing what's changing.

Not to mention something like wind up/down that _also_ have speed adjustments but don't get any special hotkeys.

bdach commented on osu.Game/Screens/Select/SongSelect.cs:
```diff
@@ -1017,6 +1109,14 @@ public virtual bool OnPressed(KeyBindingPressEvent<GlobalAction> e)
                     FinaliseSelection();
                     return true;
+
+                case GlobalAction.IncreaseSpeed:
+                    increaseSpeed();
+                    return true;
+
+                case GlobalAction.DecreaseSpeed:
+                    decreaseSpeed();
+                    return true;
```
`increaseSpeed()` and `decreaseSpeed()` should probably be one method that accept a delta or something.

bdach commented on osu.Game/Screens/Select/SongSelect.cs:
```diff
@@ -809,6 +809,98 @@ public override bool OnBackButton()
 
+        private void increaseSpeed()
+        {
+            var rateAdjustStates = ModSelect.AllAvailableMods.Where(pair => pair.Mod is ModRateAdjust);
+            var stateDoubleTime = ModSelect.AllAvailableMods.First(pair => pair.Mod is ModDoubleTime);
+            bool rateModActive = ModSelect.AllAvailableMods.Count(pair => pair.Mod is ModRateAdjust && pair.Active.Value) > 0;
```
These should not be doing mod queries or mutations via `ModSelect`. Use `selectedMods` and `OsuGameBase.AvailableMods` instead.

Fabiano1337 replied:
Done in [3fdbd73](https://github.com/ppy/osu/pull/28071/commits/3fdbd735ce063653bb92b7570df716e700a9b529)

bdach commented on osu.Game/Screens/Select/SongSelect.cs:
```diff
@@ -809,6 +809,98 @@ public override bool OnBackButton()
+            }
+
+            // Find current active rateAdjust mod and modify speed, enable DoubleTime if necessary
+            foreach (var state in rateAdjustStates)
```
This should be asserting that there is only one rate adjust mod present rather than doing some weird loop.

Fabiano1337 replied:
Done in [3fdbd73](https://github.com/ppy/osu/pull/28071/commits/3fdbd735ce063653bb92b7570df716e700a9b529)

bdach reviewed (changes requested):
Missing unit tests

Fabiano1337 commented:
I created a Discussion about this feature.
https://github.com/ppy/osu/discussions/28085

bdach commented:
I guess the discussion seems to have gathered some support, but I'd still expect the review above to be addressed for this before this can be considered further.

Fabiano1337 commented:
Sure,
since i still have school and stuff going on it could take me maybe a week or two.
Im currently trying to enable the mods without using ModSelect, i can't just new mods to selectedMods since thats a readOnlyList right ?

bdach commented:
> i can't just add new mods to selectedMods since thats a readOnlyList right ?

You can replace the entire list.

Commit:
Change Inputkeys to Ctrl+Up/Ctrl+Down

Commit:
Speedchange now also works in Modselect

Commit:
Add Unit Tests

Commit:
change to single Function,
Nightcore now switches into Daycore, keep Adjustpitch after change

Fabiano1337 commented:
- Do we add support for holding the button (for example every 500ms speed increases) or do we keep it like it is now that you have to press the key again ?
- You mentioned a toast showing that something is changing, do you have any idea on what that could look like ?

bdach commented:
> * Do we add support for holding the button (for example every 500ms speed increases) or do we keep it like it is now that you have to press the key again ?

This would already be supported if not for this early return:

https://github.com/ppy/osu/blob/3fdbd735ce063653bb92b7570df716e700a9b529/osu.Game/Screens/Select/SongSelect.cs#L1138-L1139

> * You mentioned a toast showing that something is changing, do you have any idea on what that could look like ?

Use [the existing one](https://github.com/ppy/osu/blob/dedc2d2aa0bf40045a5abd8830e3402433bb44ac/osu.Game/Overlays/OSD/Toast.cs#L17). See usages elsewhere on how to display it.

Commit:
Change Speedchange behaviour to keep changing while holding key, Add Toast to nofity user what just happend

Commit:
Toast now only shows when speed is actually changed

Fabiano1337 commented:
Seems like the toast implementation 148afd1 makes the test fail, locally they still run fine on my side

peppy committed:
Basic cleanup

Before I gave up on attempting to fix the method.

peppy commented on osu.Game/Screens/Select/SongSelect.cs:
```diff
@@ -809,6 +823,125 @@ public override bool OnBackButton()
 
+        public void ChangeSpeed(double delta)
+        {
+            ModNightcore modNc = (ModNightcore)((MultiMod)game.AvailableMods.Value[ModType.DifficultyIncrease].First(mod => mod is MultiMod multiMod && multiMod.Mods.Count(modType => modType is ModNightcore) > 0)).Mods.First(mod => mod is ModNightcore);
```
This method needs a rewrite.

Fabiano1337 replied:
Could you tell me what needs to change for the rewrite so i don't do the same mistake again? (more readable, more compact, ...)

peppy committed:
Fix enum not being at end (and adjust naming)

peppy committed:
Merge branch 'master' into lazer-speedkeys

peppy commented on osu.Game/Overlays/OSD/SpeedChangeToast.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public partial class SpeedChangeToast : Toast
+    {
+        public SpeedChangeToast(OsuConfigManager config, double delta)
```
This probably should actually show the speed?

peppy committed:
Use null conditional rather than implicit not-null

Commit:
Add speed value to Toast

Commit:
Fix test not running due to floating point number inaccuacy

Commit:
Improve readability

bdach committed:
Rewrite test

bdach committed:
Rewrite implementation

bdach committed:
Fix toast strings

bdach committed:
Fix speed controls in mod select overlay not handling repeat

bdach committed:
Actually use return value

bdach committed:
Move mod speed hotkey handler to user mod select overlay

The very base class is the wrong place for it because
`FreeModSelectOverlay` inherits from it, and that one has different
assumptions and rules than "user" selection. In particular, in non-user
selection, more than one rate adjust mod may be active, which breaks the
mod speed hotkey's basic assumptions.

bdach committed:
Merge branch 'master' into lazer-speedkeys

peppy commented:
I'm pretty happy with how this feature works (very cool) and now the code is sane. Let's 🚢 it.
